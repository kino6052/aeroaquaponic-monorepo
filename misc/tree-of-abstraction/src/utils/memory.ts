export const scoreFunction = (score: number) => 1 / ((score || 1) ** 0.9 + 1);

export let randomNumber = { value: Math.random() };
