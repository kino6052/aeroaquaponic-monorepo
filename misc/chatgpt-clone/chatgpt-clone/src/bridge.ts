import { Decoupler } from "../../../../utils/Decoupler";

export const decoupler = new Decoupler({}, (state: {}) => ({}));
