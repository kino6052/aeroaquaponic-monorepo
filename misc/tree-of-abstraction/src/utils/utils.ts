import { useEffect, useState } from "react";
import { BehaviorSubject } from "rxjs";
import { IAct, Id, initialState, IState as IState } from "../bridge";
import { act } from "../services/main.service";
import { Shortcut } from "../services/shortcuts.service";
import { IEvent } from "./EventWrapper";

export const useSharedState = <T>(
  subject: BehaviorSubject<T>
): [T, (state: T) => void] => {
  const [value, setState] = useState<T>(subject.getValue());
  useEffect(() => {
    const sub = subject.subscribe((s) => setState(s));
    return () => sub.unsubscribe();
  }, [subject]);
  const newSetState = (state: T) => subject.next(state);
  return [value, newSetState];
};

export const setPartial = <T>(
  subject: BehaviorSubject<T>,
  partial: Partial<T>
) => {
  const prev = subject.getValue();
  subject.next({ ...prev, ...partial });
};

const getRandomNumbers = (length: number) => {
  const value = Array.from(
    Math.round(Math.random() * Math.pow(10, length)).toString()
  ).reverse();
  return new Array(length)
    .fill("0")
    .map((v, i) => value[i] || v)
    .reverse()
    .join("");
};

const generateId = (amount: number = 4, length: number = 4) =>
  new Array(amount)
    .fill(0)
    .map((a, i, b) => `${i && "-"}${getRandomNumbers(length)}`)
    .join("");

export const Utils = {
  generateId,
};

export const compose = <T>(
  arr: Array<[Shortcut, (state: T, event: IEvent) => T | false]>
) => (state: T, event: IEvent): T | false =>
  arr.reduce((acc, [shortcut, cb]) => {
    const result =
      event[0] === "keydown" &&
      event[1] === Id.Keyboard &&
      event[2] === shortcut &&
      cb(state, event);
    return acc || result;
  }, false as T | false);

export const genericSequence = <T>(act: IAct<T>, initialState: T) => (
  inputs: IEvent[]
): T => {
  return inputs.reduce((acc, input) => act(acc)(input), initialState);
};
