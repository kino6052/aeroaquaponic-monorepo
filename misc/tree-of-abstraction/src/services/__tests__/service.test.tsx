import { Id, initialState } from "../../bridge";
import { Utils } from "../../utils/utils";
import { act, sequence } from "../service";

let counter = 0;

beforeAll(() => {
  jest.spyOn(Utils, "generateId").mockImplementation(() => {
    const prev = counter;
    counter += 1;
    return `${prev}`;
  });
});

beforeEach(() => {
  counter = 0;
});

afterEach(() => {
  jest.clearAllMocks();
});

afterAll(() => {
  jest.restoreAllMocks();
});

describe("App", () => {
  it("should act", () => {
    expect(act(initialState)(["click", "", ""])).toMatchInlineSnapshot(`
      Object {
        "addItemInput": "",
        "itemSearchInput": "",
        "selectedNode": "",
        "tree": Object {},
        "treeNodes": Object {
          "root": Object {
            "children": Array [],
            "id": "root",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "",
            "title": "ROOT",
          },
        },
      }
    `);
  });
});

describe("Tree", () => {
  it("should add node", () => {
    expect(
      sequence([
        ["change", Id.AddItemInput, "one"],
        ["click", Id.AddItemButton, ""],
      ])
    ).toMatchInlineSnapshot(`
      Object {
        "addItemInput": "one",
        "itemSearchInput": "",
        "selectedNode": "",
        "tree": Object {
          "root": Object {
            "item-0": Object {},
          },
        },
        "treeNodes": Object {
          "item-0": Object {
            "children": Array [],
            "id": "item-0",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "root",
            "title": "one",
          },
          "root": Object {
            "children": Array [
              "item-0",
            ],
            "id": "root",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "",
            "title": "ROOT",
          },
        },
      }
    `);
  });

  it("should add node", () => {
    expect(
      sequence([
        ["change", Id.AddItemInput, "one"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-0`, ""],
        ["change", Id.AddItemInput, "two"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-0`, ""],
        ["change", Id.AddItemInput, "three"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-2`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
      ])
    ).toMatchInlineSnapshot(`
      Object {
        "addItemInput": "four",
        "itemSearchInput": "",
        "selectedNode": "item-3",
        "tree": Object {
          "root": Object {
            "item-0": Object {
              "item-1": Object {},
              "item-2": Object {
                "item-3": Object {
                  "item-4": Object {},
                  "item-5": Object {},
                  "item-6": Object {},
                },
              },
            },
          },
        },
        "treeNodes": Object {
          "item-0": Object {
            "children": Array [
              "item-1",
              "item-2",
            ],
            "id": "item-0",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "root",
            "title": "one",
          },
          "item-1": Object {
            "children": Array [],
            "id": "item-1",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-0",
            "title": "two",
          },
          "item-2": Object {
            "children": Array [
              "item-3",
            ],
            "id": "item-2",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-0",
            "title": "three",
          },
          "item-3": Object {
            "children": Array [
              "item-4",
              "item-5",
              "item-6",
            ],
            "id": "item-3",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-2",
            "title": "four",
          },
          "item-4": Object {
            "children": Array [],
            "id": "item-4",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-3",
            "title": "four",
          },
          "item-5": Object {
            "children": Array [],
            "id": "item-5",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-3",
            "title": "four",
          },
          "item-6": Object {
            "children": Array [],
            "id": "item-6",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-3",
            "title": "four",
          },
          "root": Object {
            "children": Array [
              "item-0",
            ],
            "id": "root",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "",
            "title": "ROOT",
          },
        },
      }
    `);
  });

  it("should search", () => {
    expect(
      sequence([
        ["change", Id.AddItemInput, "one"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-0`, ""],
        ["change", Id.AddItemInput, "two"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-0`, ""],
        ["change", Id.AddItemInput, "three"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-2`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.SearchItemsInput, "o"],
      ])
    ).toMatchInlineSnapshot(`
      Object {
        "addItemInput": "four",
        "itemSearchInput": "o",
        "selectedNode": "item-3",
        "tree": Object {
          "root": Object {
            "item-0": Object {
              "item-1": Object {},
              "item-2": Object {
                "item-3": Object {
                  "item-4": Object {},
                  "item-5": Object {},
                  "item-6": Object {},
                },
              },
            },
          },
        },
        "treeNodes": Object {
          "item-0": Object {
            "children": Array [
              "item-1",
              "item-2",
            ],
            "id": "item-0",
            "isCollapsed": false,
            "isHighlighted": true,
            "parent": "root",
            "title": "one",
          },
          "item-1": Object {
            "children": Array [],
            "id": "item-1",
            "isCollapsed": false,
            "isHighlighted": true,
            "parent": "item-0",
            "title": "two",
          },
          "item-2": Object {
            "children": Array [
              "item-3",
            ],
            "id": "item-2",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-0",
            "title": "three",
          },
          "item-3": Object {
            "children": Array [
              "item-4",
              "item-5",
              "item-6",
            ],
            "id": "item-3",
            "isCollapsed": false,
            "isHighlighted": true,
            "parent": "item-2",
            "title": "four",
          },
          "item-4": Object {
            "children": Array [],
            "id": "item-4",
            "isCollapsed": false,
            "isHighlighted": true,
            "parent": "item-3",
            "title": "four",
          },
          "item-5": Object {
            "children": Array [],
            "id": "item-5",
            "isCollapsed": false,
            "isHighlighted": true,
            "parent": "item-3",
            "title": "four",
          },
          "item-6": Object {
            "children": Array [],
            "id": "item-6",
            "isCollapsed": false,
            "isHighlighted": true,
            "parent": "item-3",
            "title": "four",
          },
          "root": Object {
            "children": Array [
              "item-0",
            ],
            "id": "root",
            "isCollapsed": false,
            "isHighlighted": true,
            "parent": "",
            "title": "ROOT",
          },
        },
      }
    `);
  });

  it("should search", () => {
    expect(
      sequence([
        ["change", Id.AddItemInput, "one"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-0`, ""],
        ["change", Id.AddItemInput, "two"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-0`, ""],
        ["change", Id.AddItemInput, "three"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-2`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.SearchItemsInput, "ur"],
      ])
    ).toMatchInlineSnapshot(`
      Object {
        "addItemInput": "four",
        "itemSearchInput": "ur",
        "selectedNode": "item-3",
        "tree": Object {
          "root": Object {
            "item-0": Object {
              "item-2": Object {
                "item-3": Object {
                  "item-4": Object {},
                  "item-5": Object {},
                  "item-6": Object {},
                },
              },
            },
          },
        },
        "treeNodes": Object {
          "item-0": Object {
            "children": Array [
              "item-1",
              "item-2",
            ],
            "id": "item-0",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "root",
            "title": "one",
          },
          "item-1": Object {
            "children": Array [],
            "id": "item-1",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-0",
            "title": "two",
          },
          "item-2": Object {
            "children": Array [
              "item-3",
            ],
            "id": "item-2",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-0",
            "title": "three",
          },
          "item-3": Object {
            "children": Array [
              "item-4",
              "item-5",
              "item-6",
            ],
            "id": "item-3",
            "isCollapsed": false,
            "isHighlighted": true,
            "parent": "item-2",
            "title": "four",
          },
          "item-4": Object {
            "children": Array [],
            "id": "item-4",
            "isCollapsed": false,
            "isHighlighted": true,
            "parent": "item-3",
            "title": "four",
          },
          "item-5": Object {
            "children": Array [],
            "id": "item-5",
            "isCollapsed": false,
            "isHighlighted": true,
            "parent": "item-3",
            "title": "four",
          },
          "item-6": Object {
            "children": Array [],
            "id": "item-6",
            "isCollapsed": false,
            "isHighlighted": true,
            "parent": "item-3",
            "title": "four",
          },
          "root": Object {
            "children": Array [
              "item-0",
            ],
            "id": "root",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "",
            "title": "ROOT",
          },
        },
      }
    `);
  });

  it("should search", () => {
    expect(
      sequence([
        ["change", Id.AddItemInput, "one"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-0`, ""],
        ["change", Id.AddItemInput, "two"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-0`, ""],
        ["change", Id.AddItemInput, "three"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-2`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.SearchItemsInput, "TWO"],
      ])
    ).toMatchInlineSnapshot(`
      Object {
        "addItemInput": "four",
        "itemSearchInput": "TWO",
        "selectedNode": "item-3",
        "tree": Object {
          "root": Object {
            "item-0": Object {
              "item-1": Object {},
            },
          },
        },
        "treeNodes": Object {
          "item-0": Object {
            "children": Array [
              "item-1",
              "item-2",
            ],
            "id": "item-0",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "root",
            "title": "one",
          },
          "item-1": Object {
            "children": Array [],
            "id": "item-1",
            "isCollapsed": false,
            "isHighlighted": true,
            "parent": "item-0",
            "title": "two",
          },
          "item-2": Object {
            "children": Array [
              "item-3",
            ],
            "id": "item-2",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-0",
            "title": "three",
          },
          "item-3": Object {
            "children": Array [
              "item-4",
              "item-5",
              "item-6",
            ],
            "id": "item-3",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-2",
            "title": "four",
          },
          "item-4": Object {
            "children": Array [],
            "id": "item-4",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-3",
            "title": "four",
          },
          "item-5": Object {
            "children": Array [],
            "id": "item-5",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-3",
            "title": "four",
          },
          "item-6": Object {
            "children": Array [],
            "id": "item-6",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-3",
            "title": "four",
          },
          "root": Object {
            "children": Array [
              "item-0",
            ],
            "id": "root",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "",
            "title": "ROOT",
          },
        },
      }
    `);
  });

  it("should search", () => {
    expect(
      sequence([
        ["change", Id.AddItemInput, "one"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-0`, ""],
        ["change", Id.AddItemInput, "two"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-0`, ""],
        ["change", Id.AddItemInput, "three"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-2`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.SearchItemsInput, ""],
      ])
    ).toMatchInlineSnapshot(`
      Object {
        "addItemInput": "four",
        "itemSearchInput": "",
        "selectedNode": "item-3",
        "tree": Object {
          "root": Object {
            "item-0": Object {
              "item-1": Object {},
              "item-2": Object {
                "item-3": Object {
                  "item-4": Object {},
                  "item-5": Object {},
                  "item-6": Object {},
                },
              },
            },
          },
        },
        "treeNodes": Object {
          "item-0": Object {
            "children": Array [
              "item-1",
              "item-2",
            ],
            "id": "item-0",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "root",
            "title": "one",
          },
          "item-1": Object {
            "children": Array [],
            "id": "item-1",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-0",
            "title": "two",
          },
          "item-2": Object {
            "children": Array [
              "item-3",
            ],
            "id": "item-2",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-0",
            "title": "three",
          },
          "item-3": Object {
            "children": Array [
              "item-4",
              "item-5",
              "item-6",
            ],
            "id": "item-3",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-2",
            "title": "four",
          },
          "item-4": Object {
            "children": Array [],
            "id": "item-4",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-3",
            "title": "four",
          },
          "item-5": Object {
            "children": Array [],
            "id": "item-5",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-3",
            "title": "four",
          },
          "item-6": Object {
            "children": Array [],
            "id": "item-6",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-3",
            "title": "four",
          },
          "root": Object {
            "children": Array [
              "item-0",
            ],
            "id": "root",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "",
            "title": "ROOT",
          },
        },
      }
    `);
  });

  it("should collapse", () => {
    counter = 0;
    expect(
      sequence([
        ["change", Id.AddItemInput, "one"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-0`, ""],
        ["change", Id.AddItemInput, "two"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-0`, ""],
        ["change", Id.AddItemInput, "three"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-2`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["change", Id.AddItemInput, "four"],
        ["click", Id.AddItemButton, ""],
        ["click", `${Id.Item}-3`, ""],
        ["click", `${Id.CollapseItemButton}-2`, ""],
      ])
    ).toMatchInlineSnapshot(`
      Object {
        "addItemInput": "four",
        "itemSearchInput": "",
        "selectedNode": "item-3",
        "tree": Object {
          "root": Object {
            "item-0": Object {
              "item-1": Object {},
              "item-2": Object {},
            },
          },
        },
        "treeNodes": Object {
          "item-0": Object {
            "children": Array [
              "item-1",
              "item-2",
            ],
            "id": "item-0",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "root",
            "title": "one",
          },
          "item-1": Object {
            "children": Array [],
            "id": "item-1",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-0",
            "title": "two",
          },
          "item-2": Object {
            "children": Array [
              "item-3",
            ],
            "id": "item-2",
            "isCollapsed": true,
            "isHighlighted": false,
            "parent": "item-0",
            "title": "three",
          },
          "item-3": Object {
            "children": Array [
              "item-4",
              "item-5",
              "item-6",
            ],
            "id": "item-3",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-2",
            "title": "four",
          },
          "item-4": Object {
            "children": Array [],
            "id": "item-4",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-3",
            "title": "four",
          },
          "item-5": Object {
            "children": Array [],
            "id": "item-5",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-3",
            "title": "four",
          },
          "item-6": Object {
            "children": Array [],
            "id": "item-6",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "item-3",
            "title": "four",
          },
          "root": Object {
            "children": Array [
              "item-0",
            ],
            "id": "root",
            "isCollapsed": false,
            "isHighlighted": false,
            "parent": "",
            "title": "ROOT",
          },
        },
      }
    `);
  });
});
