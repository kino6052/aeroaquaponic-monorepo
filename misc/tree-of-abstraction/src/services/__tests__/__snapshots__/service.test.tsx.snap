// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should act 1`] = `
Object {
  "addItemInput": "",
  "currentId": 0,
  "itemSearchInput": "",
  "selectedNode": "",
  "tree": Object {},
  "treeNodes": Object {},
}
`;

exports[`Tree should add node 1`] = `
Object {
  "addItemInput": "one",
  "currentId": 1,
  "itemSearchInput": "",
  "selectedNode": "",
  "tree": Object {
    "item-0": Object {},
  },
  "treeNodes": Object {
    "item-0": Object {
      "children": Array [],
      "id": "item-0",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "",
      "title": "one",
    },
  },
}
`;

exports[`Tree should add node 2`] = `
Object {
  "addItemInput": "four",
  "currentId": 7,
  "itemSearchInput": "",
  "selectedNode": "item-3",
  "tree": Object {
    "item-0": Object {
      "item-1": Object {},
      "item-2": Object {
        "item-3": Object {
          "item-4": Object {},
          "item-5": Object {},
          "item-6": Object {},
        },
      },
    },
  },
  "treeNodes": Object {
    "item-0": Object {
      "children": Array [
        "item-1",
        "item-2",
      ],
      "id": "item-0",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "",
      "title": "one",
    },
    "item-1": Object {
      "children": Array [],
      "id": "item-1",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-0",
      "title": "two",
    },
    "item-2": Object {
      "children": Array [
        "item-3",
      ],
      "id": "item-2",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-0",
      "title": "three",
    },
    "item-3": Object {
      "children": Array [
        "item-4",
        "item-5",
        "item-6",
      ],
      "id": "item-3",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-2",
      "title": "four",
    },
    "item-4": Object {
      "children": Array [],
      "id": "item-4",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-3",
      "title": "four",
    },
    "item-5": Object {
      "children": Array [],
      "id": "item-5",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-3",
      "title": "four",
    },
    "item-6": Object {
      "children": Array [],
      "id": "item-6",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-3",
      "title": "four",
    },
  },
}
`;

exports[`Tree should collapse 1`] = `
Object {
  "addItemInput": "four",
  "currentId": 7,
  "itemSearchInput": "",
  "selectedNode": "item-3",
  "tree": Object {
    "item-0": Object {
      "item-1": Object {},
      "item-2": Object {},
    },
  },
  "treeNodes": Object {
    "item-0": Object {
      "children": Array [
        "item-1",
        "item-2",
      ],
      "id": "item-0",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "",
      "title": "one",
    },
    "item-1": Object {
      "children": Array [],
      "id": "item-1",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-0",
      "title": "two",
    },
    "item-2": Object {
      "children": Array [
        "item-3",
      ],
      "id": "item-2",
      "isCollapsed": true,
      "isHighlighted": false,
      "parent": "item-0",
      "title": "three",
    },
    "item-3": Object {
      "children": Array [
        "item-4",
        "item-5",
        "item-6",
      ],
      "id": "item-3",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-2",
      "title": "four",
    },
    "item-4": Object {
      "children": Array [],
      "id": "item-4",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-3",
      "title": "four",
    },
    "item-5": Object {
      "children": Array [],
      "id": "item-5",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-3",
      "title": "four",
    },
    "item-6": Object {
      "children": Array [],
      "id": "item-6",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-3",
      "title": "four",
    },
  },
}
`;

exports[`Tree should search 1`] = `
Object {
  "addItemInput": "four",
  "currentId": 7,
  "itemSearchInput": "o",
  "selectedNode": "item-3",
  "tree": Object {
    "item-0": Object {
      "item-1": Object {},
      "item-2": Object {
        "item-3": Object {
          "item-4": Object {},
          "item-5": Object {},
          "item-6": Object {},
        },
      },
    },
  },
  "treeNodes": Object {
    "item-0": Object {
      "children": Array [
        "item-1",
        "item-2",
      ],
      "id": "item-0",
      "isCollapsed": false,
      "isHighlighted": true,
      "parent": "",
      "title": "one",
    },
    "item-1": Object {
      "children": Array [],
      "id": "item-1",
      "isCollapsed": false,
      "isHighlighted": true,
      "parent": "item-0",
      "title": "two",
    },
    "item-2": Object {
      "children": Array [
        "item-3",
      ],
      "id": "item-2",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-0",
      "title": "three",
    },
    "item-3": Object {
      "children": Array [
        "item-4",
        "item-5",
        "item-6",
      ],
      "id": "item-3",
      "isCollapsed": false,
      "isHighlighted": true,
      "parent": "item-2",
      "title": "four",
    },
    "item-4": Object {
      "children": Array [],
      "id": "item-4",
      "isCollapsed": false,
      "isHighlighted": true,
      "parent": "item-3",
      "title": "four",
    },
    "item-5": Object {
      "children": Array [],
      "id": "item-5",
      "isCollapsed": false,
      "isHighlighted": true,
      "parent": "item-3",
      "title": "four",
    },
    "item-6": Object {
      "children": Array [],
      "id": "item-6",
      "isCollapsed": false,
      "isHighlighted": true,
      "parent": "item-3",
      "title": "four",
    },
  },
}
`;

exports[`Tree should search 2`] = `
Object {
  "addItemInput": "four",
  "currentId": 7,
  "itemSearchInput": "ur",
  "selectedNode": "item-3",
  "tree": Object {
    "item-0": Object {
      "item-2": Object {
        "item-3": Object {
          "item-4": Object {},
          "item-5": Object {},
          "item-6": Object {},
        },
      },
    },
  },
  "treeNodes": Object {
    "item-0": Object {
      "children": Array [
        "item-1",
        "item-2",
      ],
      "id": "item-0",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "",
      "title": "one",
    },
    "item-1": Object {
      "children": Array [],
      "id": "item-1",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-0",
      "title": "two",
    },
    "item-2": Object {
      "children": Array [
        "item-3",
      ],
      "id": "item-2",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-0",
      "title": "three",
    },
    "item-3": Object {
      "children": Array [
        "item-4",
        "item-5",
        "item-6",
      ],
      "id": "item-3",
      "isCollapsed": false,
      "isHighlighted": true,
      "parent": "item-2",
      "title": "four",
    },
    "item-4": Object {
      "children": Array [],
      "id": "item-4",
      "isCollapsed": false,
      "isHighlighted": true,
      "parent": "item-3",
      "title": "four",
    },
    "item-5": Object {
      "children": Array [],
      "id": "item-5",
      "isCollapsed": false,
      "isHighlighted": true,
      "parent": "item-3",
      "title": "four",
    },
    "item-6": Object {
      "children": Array [],
      "id": "item-6",
      "isCollapsed": false,
      "isHighlighted": true,
      "parent": "item-3",
      "title": "four",
    },
  },
}
`;

exports[`Tree should search 3`] = `
Object {
  "addItemInput": "four",
  "currentId": 7,
  "itemSearchInput": "TWO",
  "selectedNode": "item-3",
  "tree": Object {
    "item-0": Object {
      "item-1": Object {},
    },
  },
  "treeNodes": Object {
    "item-0": Object {
      "children": Array [
        "item-1",
        "item-2",
      ],
      "id": "item-0",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "",
      "title": "one",
    },
    "item-1": Object {
      "children": Array [],
      "id": "item-1",
      "isCollapsed": false,
      "isHighlighted": true,
      "parent": "item-0",
      "title": "two",
    },
    "item-2": Object {
      "children": Array [
        "item-3",
      ],
      "id": "item-2",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-0",
      "title": "three",
    },
    "item-3": Object {
      "children": Array [
        "item-4",
        "item-5",
        "item-6",
      ],
      "id": "item-3",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-2",
      "title": "four",
    },
    "item-4": Object {
      "children": Array [],
      "id": "item-4",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-3",
      "title": "four",
    },
    "item-5": Object {
      "children": Array [],
      "id": "item-5",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-3",
      "title": "four",
    },
    "item-6": Object {
      "children": Array [],
      "id": "item-6",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-3",
      "title": "four",
    },
  },
}
`;

exports[`Tree should search 4`] = `
Object {
  "addItemInput": "four",
  "currentId": 7,
  "itemSearchInput": "",
  "selectedNode": "item-3",
  "tree": Object {
    "item-0": Object {
      "item-1": Object {},
      "item-2": Object {
        "item-3": Object {
          "item-4": Object {},
          "item-5": Object {},
          "item-6": Object {},
        },
      },
    },
  },
  "treeNodes": Object {
    "item-0": Object {
      "children": Array [
        "item-1",
        "item-2",
      ],
      "id": "item-0",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "",
      "title": "one",
    },
    "item-1": Object {
      "children": Array [],
      "id": "item-1",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-0",
      "title": "two",
    },
    "item-2": Object {
      "children": Array [
        "item-3",
      ],
      "id": "item-2",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-0",
      "title": "three",
    },
    "item-3": Object {
      "children": Array [
        "item-4",
        "item-5",
        "item-6",
      ],
      "id": "item-3",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-2",
      "title": "four",
    },
    "item-4": Object {
      "children": Array [],
      "id": "item-4",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-3",
      "title": "four",
    },
    "item-5": Object {
      "children": Array [],
      "id": "item-5",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-3",
      "title": "four",
    },
    "item-6": Object {
      "children": Array [],
      "id": "item-6",
      "isCollapsed": false,
      "isHighlighted": false,
      "parent": "item-3",
      "title": "four",
    },
  },
}
`;
